WordPressKit/HTTPRequestBuilder.swift

```diff
@@ -0,0 +1,140 @@
+import Foundation
+
+final class HTTPRequestBuilder {
+    enum Method: String {
+        case get = "GET"
+        case post = "POST"
+        case put = "PUT"
+        case patch = "PATCH"
+        case delete = "DELETE"
+
+        var allowsHTTPBody: Bool {
+            self == .post || self == .put || self == .patch
+        }
+    }
+    private var urlComponents: URLComponents
+    private var method: Method = .get
+    private var path: String
+    private var headers: [String: String] = [:]
+    private var bodyBuilder: ((inout URLRequest) throws -> Void)?
+
+    init(url: URL) {
+        assert(url.scheme == "http" || url.scheme == "https")
+        assert(url.host != nil)
+
+        urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: true)!
+        path = urlComponents.path
+    }
+
+    func set(method: Method) -> Self {
+        self.method = method
+        return self
+    }
+
+    func set(path: String) -> Self {
+        assert(!path.contains("?") && !path.contains("#"), "Path should not have query or fragment: \(path)")
+
+        self.path = path
+        return self
+    }
+
+    func set(value: String?, forHeader header: String) -> Self {
+        headers[header] = value
+        return self
+    }
+
+    func query(name: String, value: String?, override: Bool = false) -> Self {
+        append(query: [URLQueryItem(name: name, value: value)], override: override)
+    }
```

Comment:
All the methods before this use "set" as their verb. 

I think if it would be good to use it consistently, to have a homogeneous usage at call site. E.g. from

```swift
var request = try HTTPRequestBuilder(url: URL(string: "https://wordpress.org/")!)
            .set(method: .post)
            .body(json: 42)
```

to

```swift
var request = try HTTPRequestBuilder(url: URL(string: "https://wordpress.org/")!)
            .set(method: .post)
            .set(jsonBody: 42)
```

or

```swift
var request = try HTTPRequestBuilder(url: URL(string: "https://wordpress.org/")!)
            .setMethod(.post)
            .setJSON(body: 42)
```

Reply:
Good catch! I took the other direction: removing the "set" part, because it's redundant considering this is a "builder" type. See https://github.com/wordpress-mobile/WordPressKit-iOS/pull/657/commits/b52e98d572230e251ee70c2d997611b18df1997d.

Reply:
Fine by me 😄